documentation/public/internals/design-docs/language-definition-v2/index.md

```diff
@@ -0,0 +1,438 @@
+to make sure that the resulting identifier doesn't match a reserved keyword, and if so, we should report an error, but
+continue parsing.
+
+Unique to Solidity, keywords can be `reserved_in` versions before or after the versions they are `enabled_in`. They can also be
+`unreserved_in` versions before or after the versions they are `disabled_in`. So we have to have these additional checks,
+to be able to catch cases like when a certain input can both be a keyword and an identifier, or neither.
+
+We should also be able to generate a public API `is_keyword(TerminalKind)` for users to conveniently detect them if needed.
+
+### Trivia Items
+
+Trivia items are similar tokens, contributing their own `TerminalKind`. They are referred to from the language's top-level
+`leading_trivia` and `trailing_trivia` properties. Before and after each token, the scanner should try to scan these tokens,
+collecting them in a flat list.
+
+Previously, we used to create many `LeadingTrivia` and `TrailingTrivia` nodes that hold whitespace/comments. Not only this
+is wasteful memory-wise, it is also unnatural/unexpected to wrap whitespace in non-terminal nodes. Instead, I propose
+treating them like any other token, and storing them as siblings to the tokens they belong to (in-order). Not only this
+is simpler, it is also more efficient, and is natural to how input is consumed and produced.
+
+We should also be able to generate public APIs `is_leading_trivia(TerminalKind)` and `is_trailing_trivia(TerminalKind)`
+for users to conveniently detect them if needed.
+
+### Fragment Items
+
+Fragments are not visible to users, and don't contribute a `TerminalKind`. They are just a utility used to refactor
+common parts of the grammar, and avoid duplication. During processing the language definition, they are inlined wherever
+they are referenced.
+
+## NonTerminals
+
+### Struct Items
+
+Structs represent a flat list (sequence) of typed fields. They are the simplest non-terminal, and generate a `struct` AST type.
+Their fields match 1-1 with the item fields. The struct name contributes a `NonTerminalKind`.
+
+Each field can be either `Required(T)` or `Optional(T)`. Required fields are always present and parsed. Optional fields
+can be omitted if they don't exist, and are represented with Rust's `Option<T>` type (or TypeScript `T | undefined`).
+However, optional fields have additional `enabled_in` and `disabled_in` properties. After parsing optional fields,
+we should compare them with the current language version, and produce errors if they don't match, but continue parsing normally.
+
+The type of each field can be a `NonTerminal(T)` or `Terminal(Set<T>)`. A non-terminal field refers to another item, and
+holds its type. A terminal field refers to one or more terminal items (all valid in this position), and is of type `TerminalNode`.
+
+Additionally, the struct also stores the CST node that holds its contents. The idea is for the parser to attempt
```

Comment:
Simultaneous construction not yet agreed upon IIRC

Reply:
Removed.